### ISSUE TITLE
Add an example to next.js

### ISSUE THREAD

**Issue by kof** [cssinjs/jss#457]
https://github.com/zeit/next.js/tree/master/examples

**Comment by jcheroske** [comment]
Is it currently possible to swap out `styled-jsx` for `jss` when using `next.js`? I was looking at zeit/styled-jsx#142 and didn't think that `jss` suffers from that kind of problem. Not being able to style component libraries without resorting to global classes seems like a major shortcoming.

**Comment by kof** [comment]
I am 99% sure you can use jss there, I haven't tried though. Please add an example over there as you go!

And yes, jss has jss-global plugin which is part of the jss-preset-default and allows to use global selectors and scoped global selectors.

**Comment by jcheroske** [comment]
Here is an issue and pull request to look at: zeit/next.js#1751 and zeit/next.js#1776

The situation is that next.js currently doesn't have any way to wrap all pages with additional components. I forked it and added the functionality myself in order to get SSR working. There are hacks available to get it working without the modification of `renderPage`, but they have no place in good documentation. I'm happy to create the example when the API changes.

I'm happy to look into fixing the SSR counter reset bug, since I currently have a working system that demonstrates it.

**Comment by jcheroske** [comment]
And for completeness, here's my comment, copied from #432:

I'm seeing this in next.js as well:

Warning: React attempted to reuse markup in a container but the checksum was invalid. This 
generally means that you are using server rendering and the markup generated on the server was
not what the client was expecting. React injected new markup to compensate which works but you
have lost many of the benefits of server rendering. Instead, figure out why the markup being 
generated is different on the client or server:
 (client) <div class="paper-0-0" style="color:rgba
 (server) <div class="paper-0-2" style="color:rgba
When the next.js server first boots, the jss counter is at 0. If I hit my page everything works, and there is no React warning. If I refresh the browser, for some reason next.js reruns my component file, and re-injects the sheet, which increments the server-side jss counter (This only happens one time). But the browser got refreshed, so its counter is back to zero. Additionally, every time the dev server detects a file change it reruns all the components, and the counter gets incremented yet again.

I think you guys are right in thinking the the load order of React components is deterministic. What might not be though is the counter's starting point on the server. SSR is a weird beast. If you have five sheets, the counter will always increment five times in the same sheet order. But, where it always runs 0, 1, 2, 3, 4 on the client, it might run 13,14,15,16,17 on the server.

And, just to make sure I'm not missing something, is there a special format for the server-generated <style> tag? Meta info of some kind? Reading this thread, I didn't think so.

**Comment by kof** [comment]
Sure, I am happy to receive a pr.

**Comment by kof** [comment]
Can you show your JSS example written for next? Maybe something is done in a wrong way…

**Comment by jcheroske** [comment]
Very likely! So, as far as I know, to get SSR working I just have to create a new `SheetsRegistry` on every request, and then pass it in via the `SheetsRegistryProvider`. What I'm not doing is creating a new `jss` instance on every request. I didn't think that was necessary, but maybe it is. Here is my `pages/_document.js`:
```
import Document, { Head, Main, NextScript } from 'next/document'
import {SheetsRegistry, SheetsRegistryProvider} from 'react-jss'
import 'next-extras'

export default class EnhancedDocument extends Document {
  static async getInitialProps ({renderPage}) {
    const sheets = new SheetsRegistry()

    const decoratePage = Page => props => (
      <SheetsRegistryProvider registry={sheets}>
        <Page {...props} />
      </SheetsRegistryProvider>
    )

    const renderedPage = renderPage(decoratePage)
    const styles = (
      <style type='text/css' data-meta='jss-ssr'>
        {sheets.toString()}
      </style>
    )
    const props = {
      ...renderedPage,
      styles
    }
    return props
  }

  render () {
    return (
      <html>
        <Head>
          <link rel='preload stylesheet' href='https://fonts.googleapis.com/css?family=Roboto:300,400,500' />
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </html>
    )
  }
}
```

**Comment by kof** [comment]
Ok its all correct, we just need to reset counters on each request.

**Comment by jcheroske** [comment]
Looking at the code, I see the following:

* Classnames are generated, by default, by `utils/generateClassName`, which has a rule counter in module scope
* The particular classname generator is linked to a `jss` instance when `jss/create` is called. I haven't seen another way to override the generator after that point.
* During SSR, each request needs its own counter. Apparently the only way to do that would be to create a new `jss` instance and then pass it into `react-jss/create`, creating a new HOC for each request. This is complicated, since code just wants to `import injectSheet` from 'react-jss'.

Before I go further, I'd like to know if the above points are correct? I'd also like to know if it's possible to have the default `jss` instance use the old generator?

**Comment by kof** [comment]
We would need to bound counter to some instance which is created on each request, I was thinking about `SheetsRegistry`, because it will be created for a new request every time.

**Comment by jcheroske** [comment]
@kof I've got this fixed locally now. It required mods to both `jss` and `react-jss`, but nothing severe. My goal was to maintain complete backwards compatibility, which meant I created an options object for `SheetsRegistry`, that, if omitted, keeps all the existing behavior. I'm putting the finishing touches on now and will push soon.

**Comment by kof** [comment]
sure, send me a pr so that we can discuss the code

**Comment by oliviertassinari** [comment]
For what that worth, [there is an example](https://github.com/zeit/next.js/tree/v3-beta/examples/with-material-ui-next) with Material-UI that is using JSS internally.

**Comment by jacobpdq** [comment]
Hey all, looking for SSR / next export functionality. Any chance I could get in on dev here to make that possible?

**Comment by kof** [comment]
changes we need in the core are done, next thing react-jss changes and a release

**Comment by jacobpdq** [comment]
OK, thanks – I'll watch react-jss

**Comment by kof** [comment]
Created a corresponding issue https://github.com/cssinjs/react-jss/issues/101

**Comment by iamstarkov** [comment]
i have wip https://github.com/iamstarkov/react-jss-next-demo

**Comment by iamstarkov** [comment]
`./pages/_document.js` looks like this:

```js
import Document, { Head, Main, NextScript } from 'next/document'
import { SheetsRegistry, SheetsRegistryProvider } from 'react-jss'

export default class EnhancedDocument extends Document {
  static async getInitialProps ({ renderPage }) {
    const sheets = new SheetsRegistry()

    const decoratedPage = Page => props => (
      <SheetsRegistryProvider registry={sheets}>
        <Page {...props} />
      </SheetsRegistryProvider>
    )

    const page = renderPage(decoratedPage)
    const ssrCss = sheets.toString();
    return { ...page, ssrCss }
  }

  render () {
    console.log(this.props)
    return (
      <html>
        <Head>
          <style data-jss dangerouslySetInnerHTML={{ __html: this.props.ssrCss }} />
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </html>
    )
  }
}

```

**Comment by iamstarkov** [comment]
but `ssrCss` is empty string =(

**Comment by iamstarkov** [comment]
@jcheroske, can you take a look?

**Comment by oliviertassinari** [comment]
@iamstarkov Might want to have a look at the hack I have done to make the Material-UI example work 🙈 .

**Comment by jcheroske** [comment]
I wrote zeit/next.js#2010 and it seems like it's going to be approved. It's going to make SSR config much easier.

**Comment by iamstarkov** [comment]
@jcheroske it would be nice to make https://github.com/zeit/next.js/pull/2010 happen, because othwerwise it doesnt work with current renderPage implementation https://github.com/zeit/next.js/blob/87e01f681bc2d82e59ee30c026180d6984157a34/server/render.js#L69-L93

**Comment by jcheroske** [comment]
Yeah, without the hook in that PR then SSR config on next.js is virtually impossible. I'm not sure what more I can do to push that PR forward. All of you guys do have clout though. If you posted in the PR thread it might help move things along.

**Comment by kof** [comment]
I think we have  a working react-jss version by now, who would like to submit it to next?

**Comment by iamstarkov** [comment]
I tried. we have two options:

1) follow very cumbersome approach found be `material-ui/next` https://github.com/zeit/next.js/blob/v3-beta/examples/with-material-ui-next/pages/_document.js
2) or wait until https://github.com/zeit/next.js/pull/2010 being merged and released

**Comment by oliviertassinari** [comment]
I have used a singleton to make it work. Not great! But as long as you render the requests sequencially, it's OK.

**Comment by kof** [comment]
@jcheroske @iamstarkov any updates?

**Comment by iamstarkov** [comment]
@kof i have working example here https://github.com/iamstarkov/react-jss-next-demo/

**Comment by kof** [comment]
@iamstarkov can you submit it to next?

**Comment by nyl-auster** [comment]
@iamstarkov thanks for your work ! I tried your example but it seems styles are added two times ( i guess they are server side rendered and then injected again in the DOM by the client ? ) 

<img width="251" alt="capture d ecran 2018-06-11 a 02 35 13" src="https://user-images.githubusercontent.com/335495/41208152-2a69f306-6d20-11e8-83d3-33374945bbcd.png">

**Comment by trusktr** [comment]
Seems like I'm having the counter reset problem (https://github.com/cssinjs/jss/issues/841).

I tried doing just like what @iamstarkov's example does, but still no luck although it made progress.

Instead of seeing an ID with a high counter like `my-class-0-1-63` I now see `my-class-0-1-13` which means the count is not going as high any more, but it is still not as low as the client side which has `my-class-0-1-3`.

**Comment by trusktr** [comment]
Hmmm, but I'm also using `jss.createStyleSheet` directly in a couple other places to set some global styles. Could that cause the problem? Maybe I should put all styles inside the React components with react-jss only?

**Comment by trusktr** [comment]
Alright, I moved _all_ styles into React components, but I still get a classname mismatch on server vs client.

**Comment by trusktr** [comment]
What about just putting the `<JssProvider>` inside the _document's `render()` method content instead of around the `Page` in getInitialProps?

**Comment by trusktr** [comment]
I'm struggling with this. I don't understand why the following log shows nothing:

```js
  static getInitialProps({ renderPage }: { renderPage(hoc: HOC): any }) {
    debugger
    let sheets: SheetsRegistry

    const decoratePage: HOC = App => props => {
      debugger
      sheets = new SheetsRegistry()

      return (
        <JssProvider registry={sheets}>
          <App {...props} />
        </JssProvider>
      )
    };

    const renderedPage = renderPage(decoratePage);

    console.log('sheets', sheets.toString()) // NOTHING <------------------------------
    debugger
    const styles = (
      <style type="text/css" data-meta="jss-ssr" dangerouslySetInnerHTML={{ __html: sheets.toString() }} />
    );

    return { ...renderedPage, styles };
  }
```

The first time the app loads it works. Refresh the page and it breaks with class name mismatch. Either way, nothing is logged to console.

**Comment by trusktr** [comment]
Okay, I deleted `getInitialProps` from `Document`, and I placed `JssProvider` in my `App`'s render method, and now it all works fine and no name clashes, but of course the styles will not be included in the initial payload now.

I have to roll with this for now, as I can't get the other way to work. And I don't have a reproduction yet because the code is private. Maybe I can make one a little later...

**Comment by trusktr** [comment]
I think I see the problem:

- I have an `_app.tsx`
- I use components inside of `_app.tsx` (for example a `<RootLayout>` component) which use react-jss' `injectStyles` (which ultimately uses a `SheetsRegistry`).
- Next.js renders the `_app.tsx` component **_before_** rendering a the route's `Page` component that we've decorated in `renderPage`.
- Therefore, the `App` component will create its own `SheetsRegistry`, and there will then be a new `SheetsRegistry` for the `Page`.

Basically apps using components styled with react-jss inside of the `App` component will encounter the class name mismatch problem due to different counter values.

I believe I can now reproduce this in the example. Let me try...

**Comment by trusktr** [comment]
Alright, here's the fix: https://github.com/iamstarkov/react-jss-next-demo/pull/5

It builds from https://github.com/iamstarkov/react-jss-next-demo/pull/4. Instead of decorating the Page component, the `_document` and `_app` now simply share a singleton `SheetsRegistry`.

---

But there's another unrelated problem. If you check the output of `sheets.toString()` you'll see it is an empty string, just like in my above comment. If you view source, you'll see

```html
      <style type="text/css" data-meta="jss-ssr"></style>
```

I think this new problem has to do with react-jss. @hof, any idea why it returns an empty string in the reproduction?

**Comment by HenriBeck** [comment]
Why does your _app also need a SheetsRegistry?

I just implemented ssr with react-jss in one of my projects and it works perfectly fine.

**Comment by trusktr** [comment]
I'm using root-level components in _app to wrap the page components with (for example) a layout which includes header and footer components. They use react-jss, therefore they need a SheetsRegistry.

TypeStyle uses a similar module-based technique as in my fix. See the example: https://typestyle.github.io/#/server

Basically we just use TypeStyle, and we don't worry about when/how to create TypeStyle internals internals. When we are ready to render the payload, we just use `getStyles`. It won't matter how many modules import and use TypeStyle, or in what order assuming.

**Comment by trusktr** [comment]
@HenriBeck It's not just for client side. The App component runs on the server too.

**Comment by trusktr** [comment]
It is working flawlessly now (it wasn't before) so I think that's worth something!

**Comment by HenriBeck** [comment]
Example was added in zeit/next.js#5140

**Comment by kof** [comment]
💥

**Comment by weilin-ong** [comment]
> Example was added in [vercel/next.js#5140](https://github.com/vercel/next.js/pull/5140)

Hi, could you please also add an example adapting next.js 13 app router? It would be very helpful for anyone who is doing migration from pages to app! thank you

**Comment by cesar3030** [comment]
Looking for an example with App Router as well!
Did you figure it out @weilin-ong ?

**Comment by weilin-ong** [comment]
> Looking for an example with App Router as well! Did you figure it out @weilin-ong ?

unfortunately no :/
